### PR TITLE
CASSANDRA-16911 - remove ephemeral marker file for snapshot

### PR DESCRIPTION
**Pull request by smiklosovic**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 1]
```diff
@@ -1698,9 +1702,9 @@ public CompactionManager.AllSSTableOpStatus verify(Verifier.Options options) thr
     /**
      * Rewrites all SSTables according to specified parameters
      *
-     * @param skipIfCurrentVersion - if {@link true}, will rewrite only SSTables that have version older than the current one ({@link BigFormat#latestVersion})
+     * @param skipIfCurrentVersion - if {@link true}, will rewrite only SSTables that have version older than the current one ({@link org.apache.cassandra.io.sstable.format.big.BigFormat#latestVersion})
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java:1702]: opportunistic bugfix, without this IDEA marks it errorneous hence all class

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 2]
```diff
@@ -1698,9 +1702,9 @@ public CompactionManager.AllSSTableOpStatus verify(Verifier.Options options) thr
...
      * @param skipIfNewerThanTimestamp - max timestamp (local creation time) for SSTable; SSTables created _after_ this timestamp will be excluded from compaction
-     * @param skipIfCompressionMatches - if {@link true}, will rewrite only SSTables whose compression parameters are different from {@link CFMetaData#compressionParams()}
+     * @param skipIfCompressionMatches - if {@link true}, will rewrite only SSTables whose compression parameters are different from {@link TableMetadata#params#getCompressionParameters()} ()}
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java:1704]: opportunistic bugfix, without this IDEA marks it errorneous hence all class

File: src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java [hunk 3]
```diff
@@ -80,43 +81,58 @@ public Collection<TableSnapshot> getExpiringSnapshots()
...
     {
         SnapshotLoader loader = new SnapshotLoader(DatabaseDescriptor.getAllDataFileLocations());
         addSnapshots(loader.loadSnapshots());
+        snapshotsLoaded = true;
     }
 
     @VisibleForTesting
     protected synchronized void addSnapshots(Collection<TableSnapshot> snapshots)
     {
-        logger.debug("Adding snapshots: {}.", Joiner.on(", ").join(snapshots.stream().map(s -> s.getId()).collect(Collectors.toList())));
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java]: I moved this to addSnapshot method where I differ between expiring and ephemeral snapshot.

File: test/unit/org/apache/cassandra/db/DirectoriesTest.java [hunk 4]
```diff
@@ -105,7 +105,7 @@
     private static Set<TableMetadata> CFM;
     private static Map<String, List<File>> sstablesByTableName;
 
-    @Parameterized.Parameter(0)
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/db/DirectoriesTest.java:108]: 0 is default

File: test/unit/org/apache/cassandra/db/DirectoriesTest.java [hunk 5]
```diff
@@ -151,15 +151,15 @@ public void beforeTest() throws IOException
     @AfterClass
     public static void afterClass()
     {
-        FileUtils.deleteRecursive(tempDataDir);
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/db/DirectoriesTest.java:154]: opportunistic fix as the used method was marked as deprecated, similar cases in this PR at various places

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 6]
```diff
@@ -817,14 +829,6 @@ public static void  scrubDataDirectories(TableMetadata metadata) throws StartupE
                     if (!file.tryDelete())
                         logger.warn("could not delete {}", file.absolutePath());
         }
-
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java]: why is this being removed?

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java (hunk 1 above)
Comment (pauloricardomg) [reply src/java/org/apache/cassandra/db/ColumnFamilyStore.java:1702]: 👍

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java (hunk 2 above)
Comment (pauloricardomg) [reply src/java/org/apache/cassandra/db/ColumnFamilyStore.java:1704]: 👍

File: src/java/org/apache/cassandra/db/Directories.java [hunk 7]
```diff
@@ -977,24 +966,30 @@ public Map<String, TableSnapshot> listSnapshots()
...
-    protected TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
+    protected TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs, boolean ephemeral) {
         Instant createdAt = manifest == null ? null : manifest.createdAt;
         Instant expiresAt = manifest == null ? null : manifest.expiresAt;
         return new TableSnapshot(metadata.keyspace, metadata.name, metadata.id.asUUID(), tag, createdAt, expiresAt,
-                                 snapshotDirs);
+                                 snapshotDirs, ephemeral);
+    }
+
+    protected static boolean isEphemeral(Set<File> snapshotDirs)
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Directories.java]: Can you add a short comment about why this method is needed - should we remove it at some point and just use the manifest to decide if a snapshot is ephemeral?

File: src/java/org/apache/cassandra/db/Directories.java [hunk 8]
```diff
@@ -1018,42 +1013,6 @@ protected static SnapshotManifest maybeLoadManifest(String keyspace, String tabl
...
-        final List<String> ephemeralSnapshots = new LinkedList<>();
-        for (File snapshot : listAllSnapshots())
-        {
-            if (getEphemeralSnapshotMarkerFile(snapshot).exists())
-                ephemeralSnapshots.add(snapshot.name());
-        }
-        return ephemeralSnapshots;
-    }
-
-    private List<File> listAllSnapshots()
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Directories.java:1032]: why is this removed?

**Review by pauloricardomg — COMMENTED** [review]
adding some initial comments (have not finished review yet)

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java (hunk 6 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/db/ColumnFamilyStore.java]: can return back

File: src/java/org/apache/cassandra/db/Directories.java (hunk 7 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/db/Directories.java]: ok i ll add comment

File: src/java/org/apache/cassandra/db/Directories.java (hunk 8 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/db/Directories.java:1032]: I think it is because that method is not necessary anymore. That method was used in `listEphemeralSnapshots` but we do not need that method anymore.

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java (hunk 6 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/db/ColumnFamilyStore.java]: ah right I see what you ask, like why I removed that block of the code? Because it is not necessary as I moved the logic up into scrubDataDirectories where I am fitlering indexes and we will scrub it with all tables as well.

File: src/java/org/apache/cassandra/db/Directories.java (hunk 7 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/db/Directories.java]: I rewrote it so that method is not needed anymore.

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java (hunk 6 above)
Comment (pauloricardomg) [reply src/java/org/apache/cassandra/db/ColumnFamilyStore.java]: ok thanks for clarifying

File: src/java/org/apache/cassandra/db/Directories.java [hunk 9]
```diff
@@ -983,18 +972,23 @@ public Map<String, TableSnapshot> listSnapshots()
         return snapshots;
     }
 
-    protected TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
+    private TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
+        boolean ephemeral;
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Directories.java]: I think a ternary if would be slightly more readable:
`boolean ephemeral = manifest != null? manifest.isEphemeral() : isLegacyEphemeralSnapshot(snapshotDirs)`

This allow us to test the `isLegacyEphemeralSnapshot(snapshotDirs)` logic on a unit test more easily since it's a static method

File: src/java/org/apache/cassandra/db/Directories.java (hunk 8 above)
Comment (pauloricardomg) [reply src/java/org/apache/cassandra/db/Directories.java:1032]: ok thanks for clarifying.

File: src/java/org/apache/cassandra/tools/nodetool/ListSnapshots.java [hunk 10]
```diff
@@ -50,6 +55,7 @@ public void execute(NodeProbe probe)
 
             Map<String, String> options = new HashMap<>();
             options.put("no_ttl", Boolean.toString(noTTL));
+            options.put("no_ephemeral", Boolean.toString(!ephemeral));
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/tools/nodetool/ListSnapshots.java]: It doesn't feel natural passing `!ephemeral`. Since the default is to **NOT** show snapshots, the property name should be called `include_ephemeral`, which defaults to false - and we can pass the `ephemeral` field without negating it.

File: test/unit/org/apache/cassandra/db/DirectoriesTest.java (hunk 5 above)
Comment (pauloricardomg) [reply test/unit/org/apache/cassandra/db/DirectoriesTest.java:154]: sounds good.

File: src/java/org/apache/cassandra/service/CassandraDaemon.java [hunk 11]
```diff
@@ -373,6 +374,10 @@ protected void setup()
             throw new RuntimeException(e);
         }
 
+        // here we preload all snapshots so once we execute SnapshotManager#start we will have all snapshots scanned already
+        StorageService.instance.snapshotManager.loadSnapshots();
+        StorageService.instance.snapshotManager.clearSnapshots(StorageService.instance.snapshotManager.getEphemeralSnapshots());
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/CassandraDaemon.java]: I didn't understand why are we moving snapshot cleaning from `scrubDataDirectories`. I thought startup leftover cleanup logic was centralized there - this will make the logic be spread around multiple places. Not sure if it's a problem or not.

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 12]
```diff
@@ -4136,6 +4136,7 @@ public void clearSnapshot(String tag, String... keyspaceNames) throws IOExceptio
     public Map<String, TabularData> getSnapshotDetails(Map<String, String> options)
     {
         boolean skipExpiring = options != null && Boolean.parseBoolean(options.getOrDefault("no_ttl", "false"));
+        boolean skipEphemeral = options != null && Boolean.parseBoolean(options.getOrDefault("no_ephemeral", "true"));
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StorageService.java]: this should probably be called `includeEphemeral` to be aligned with the nodetool parameter.

File: src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java [hunk 13]
```diff
@@ -80,43 +81,58 @@ public Collection<TableSnapshot> getExpiringSnapshots()
...
-        // We currently only care about expiring snapshots
-        if (snapshot.isExpiring())
+        // We currently care only about expiring and ephemeral snapshots
+        if (snapshot.isExpiring() && !expiringSnapshots.contains(snapshot))
         {
-            logger.debug("Adding expiring snapshot {}", snapshot);
+            logger.debug("Adding expiring snapshot {}", snapshot.getId());
             expiringSnapshots.add(snapshot);
         }
+        // we are adding ephemeral snapshots only in case we are doing the initial loading
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java]: I don't like this logic because it's inconsistent depending on the node state. We should either *always* include ephemeral, or never include ephemeral. Why do we need to load ephemeral snapshots in `SnapshotManager`? Can't we use `SnapshotLoader` to load ephemeral snapshots on `scrubDataDirectories` - no need to load on `SnapshotManager` ?

Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java]: My logic behind that was that when we put this in scrub method, we walk the disk once and then we need to walk it for the second time when we are loading snapshots in manager. I do not know if this is a big problem to scan it twice or not. I was trying to do it in such a way that we need to walk it just once. If there is 10 000 snaphots I think it starts to make a difference if we walk once or twice.

File: src/java/org/apache/cassandra/service/StorageService.java (hunk 12 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/StorageService.java]: no problem, I was trying to follow same logic we did in skipExpiring / ttl but if it reads better, why not.

File: src/java/org/apache/cassandra/service/CassandraDaemon.java (hunk 11 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/CassandraDaemon.java]: Answered below.

File: src/java/org/apache/cassandra/tools/nodetool/ListSnapshots.java (hunk 10 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/tools/nodetool/ListSnapshots.java]: ok

File: src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java (hunk 13 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java]: To add to it - the problem is that if we are going to _always_ include ephemerals, you will have them during whole life of the node because they can be removed only on startup. So we will be basically only adding snapshots. Is not this a concern from memory point of view?

Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java]: Well, I mean ... we can indeed always add it and then an operator can delete these ephemerals after he sees them via nodetool, yeah. I was not taking that possibility into account. The fact that they are _always removed on startup_ does not mean that we could not remove it while a node is up, right?

File: src/java/org/apache/cassandra/db/Directories.java [hunk 14]
```diff
@@ -973,17 +973,17 @@ public Map<String, TableSnapshot> listSnapshots()
...
-        if (manifest == null)
-            ephemeral = snapshotDirs.stream().map(d -> new File(d, "ephemeral.snapshot")).anyMatch(File::exists);
-        else
-            ephemeral = manifest.isEphemeral();
+        boolean ephemeral = manifest != null ? manifest.isEphemeral() : isLegacyEphemeralSnapshot(snapshotDirs);
         Instant createdAt = manifest == null ? null : manifest.createdAt;
         Instant expiresAt = manifest == null ? null : manifest.expiresAt;
         return new TableSnapshot(metadata.keyspace, metadata.name, metadata.id.asUUID(), tag, createdAt, expiresAt,
                                  snapshotDirs, ephemeral);
     }
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/Directories.java:983]: TODO Ideally add VisibleForTesting and test it.

File: src/java/org/apache/cassandra/db/Keyspace.java [hunk 15]
```diff
@@ -307,6 +308,9 @@ public static void clearSnapshot(String snapshotName, String keyspace)
         RateLimiter clearSnapshotRateLimiter = DatabaseDescriptor.getSnapshotRateLimiter();
 
         List<File> tableDirectories = Directories.getKSChildDirectories(keyspace);
+        if (StorageService.instance.snapshotManager.isEphemeral(snapshotName)) {
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/Keyspace.java]: This expects that we are holding ephemerals too.

File: src/java/org/apache/cassandra/service/ActiveRepairService.java [hunk 16]
```diff
@@ -786,7 +781,7 @@ public synchronized ParentRepairSession removeParentRepairSession(TimeUUID paren
                 for (ColumnFamilyStore cfs : session.columnFamilyStores.values())
                 {
                     if (cfs.snapshotExists(snapshotName))
-                        cfs.clearSnapshot(snapshotName);
+                        StorageService.instance.snapshotManager.clearEphemeralSnapshot(snapshotName);
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/service/ActiveRepairService.java]: If we hold ephemerals, we need to remove them from manager when repair wants to clear it.

File: src/java/org/apache/cassandra/service/CassandraDaemon.java [hunk 17]
```diff
@@ -292,6 +292,9 @@ protected void setup()
 
         SSTableHeaderFix.fixNonFrozenUDTIfUpgradeFrom30();
 
+        StorageService.instance.snapshotManager.loadSnapshots();
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/service/CassandraDaemon.java]: Moved before scrubbing. We are ever going to load snapshots (scanning the dirs) just once.

File: src/java/org/apache/cassandra/db/Directories.java [hunk 18]
```diff
@@ -983,18 +972,23 @@ public Map<String, TableSnapshot> listSnapshots()
...
+    private TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
+        boolean ephemeral = manifest != null ? manifest.isEphemeral() : isLegacyEphemeralSnapshot(snapshotDirs);
         Instant createdAt = manifest == null ? null : manifest.createdAt;
         Instant expiresAt = manifest == null ? null : manifest.expiresAt;
         return new TableSnapshot(metadata.keyspace, metadata.name, metadata.id.asUUID(), tag, createdAt, expiresAt,
-                                 snapshotDirs);
+                                 snapshotDirs, ephemeral);
+    }
+
+    protected static boolean isLegacyEphemeralSnapshot(Set<File> snapshotDirs) {
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/Directories.java]: this is tested as part of junits

File: src/java/org/apache/cassandra/db/Directories.java [hunk 19]
```diff
@@ -983,18 +972,23 @@ public Map<String, TableSnapshot> listSnapshots()
         return snapshots;
     }
 
-    protected TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
+    private TableSnapshot buildSnapshot(String tag, SnapshotManifest manifest, Set<File> snapshotDirs) {
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Directories.java]: nit: add line-break

File: src/java/org/apache/cassandra/db/Directories.java (hunk 18 above)
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Directories.java]: nit: add line-break

File: src/java/org/apache/cassandra/service/snapshot/SnapshotManifest.java [hunk 20]
```diff
@@ -46,19 +46,24 @@
     @JsonProperty("expires_at")
     public final Instant expiresAt;
 
+    @JsonProperty("ephemeral")
+    public final boolean ephemeral;
+
     /** needed for jackson serialization */
     @SuppressWarnings("unused")
     private SnapshotManifest() {
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/snapshot/SnapshotManifest.java]: nit: add missing line-break

File: src/java/org/apache/cassandra/service/snapshot/TableSnapshot.java [hunk 21]
```diff
@@ -239,19 +248,25 @@ void addSnapshotDir(File snapshotDir)
...
             snapshotDirs.add(snapshotDir);
             File manifestFile = new File(snapshotDir, "manifest.json");
-            if (manifestFile.exists() && createdAt == null && expiresAt == null) {
-                loadTimestampsFromManifest(manifestFile);
-            }
+            if (manifestFile.exists() && createdAt == null && expiresAt == null)
+                loadMetadataFromManifest(manifestFile);
+
+            // check if an ephemeral marker file exists only in case it is not already ephemeral
+            // by reading it from manifest
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/snapshot/TableSnapshot.java:255]: Maybe add a TODO reminder and a link to CASSANDRA-16911 to remove this on 4.3+ ?

File: test/unit/org/apache/cassandra/db/DirectoriesTest.java [hunk 22]
```diff
@@ -552,11 +552,9 @@ public void testMTSnapshots() throws Exception
             final Directories directories = new Directories(cfm, toDataDirectories(tempDataDir));
             assertEquals(cfDir(cfm), directories.getDirectoryForNewSSTables());
             final String n = Long.toString(nanoTime());
-            Callable<File> directoryGetter = new Callable<File>() {
-                public File call() throws Exception {
-                    Descriptor desc = new Descriptor(cfDir(cfm), KS, cfm.name, sstableId(1), SSTableFormat.Type.BIG);
-                    return Directories.getSnapshotDirectory(desc, n);
-                }
+            Callable<File> directoryGetter = () -> {
```
Comment (pauloricardomg) [inline test/unit/org/apache/cassandra/db/DirectoriesTest.java]: nit: add line-break

**Review by pauloricardomg — APPROVED** [review]

**Review by pauloricardomg — COMMENTED** [review]
.

**Review by pauloricardomg — CHANGES_REQUESTED** [review]
sorry approved by mistake, see comment on JIRA ticket

File: test/distributed/org/apache/cassandra/distributed/impl/Instance.java [hunk 23]
```diff
@@ -620,6 +621,19 @@ public void startup(ICluster cluster)
                 // Start up virtual table support
                 CassandraDaemon.getInstanceForTesting().setupVirtualKeyspaces();
 
+                // clean up debris in the rest of the keyspaces
+                for (String keyspaceName : Schema.instance.getKeyspaces())
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/impl/Instance.java]: this needs to be here, actually, it should be here already. if you see what this method does, it basically copies what is in CassandraDaemon setup minus some details. I followed that method closely and this is what was missing. Without this, it will not trigger the code which scrubs data dirs / removes ephemerals.

Comment (pauloricardomg) [reply test/distributed/org/apache/cassandra/distributed/impl/Instance.java]: 👍 

To avoid code duplication, do you mind extracting this common logic to `CassandraDaemon.scrubDataDirectories()`? And then we could call `CassandraDaemon.getInstanceForTesting().scrubDataDirectories()` here as done in `CassandraDaemon.getInstanceForTesting().setupVirtualKeyspaces()`.

**Review by pauloricardomg — APPROVED** [review]
